tagmanager/src/tm_workspace.c

```diff
@@ -823,514 +728,333 @@ fill_find_tags_array (GPtrArray *dst, const GPtrArray *src,
-			}
+			tag = TM_TAG(type_tags->pdata[j]);
+			/* prefer non-typedef tags because we can be sure they contain members */
+			if (tag->type != tm_tag_typedef_t)
+				break;
 		}
-	}
 
-	if (local->len > 0)
-	{
-		char *scope;
-		for (i = 0; (i < local->len); ++i)
+		g_ptr_array_free(type_tags, TRUE);
+
+		if (!tag) /* not a type that can contain members */
+			break;
+
+		/* intermediate typedef - resolve to the real type */
+		if (tag->type == tm_tag_typedef_t && tag->var_type && tag->var_type[0] != '\0')
 		{
-			tag = TM_TAG (local->pdata[i]);
-			scope = tag->scope;
-
-			/* if we wanna complete something like
-			 * namespace1::
-			 * we'll just return the tags that have "namespace1"
-			 * as their scope. So we won't return classes/members/namespaces
-			 * under, for example, namespace2, where namespace1::namespace2
-			 */
-			if (scope && 0 == strcmp (name, scope))
-			{
-				g_ptr_array_add (tags, tag);
-			}
+			type_name = tag->var_type;
+			continue;
+		}
+		else /* real type with members */
+		{
+			has_members = TRUE;
+			break;
 		}
 	}
 
-	g_ptr_array_free (local, TRUE);
-	return (int) tags->len;
-}
-
-static const GPtrArray *
-tm_workspace_find_namespace_members (const GPtrArray * file_tags, const char *name,
-								 gboolean search_global)
-{
-	static GPtrArray *tags = NULL;
-	GPtrArray *local = NULL;
-	char *new_name = (char *) name;
-	char *filename = NULL;
-	int found = 0, del = 0;
-	static langType langJava = -1;
-	TMTag *tag = NULL;
+	if (has_members)
+		/* use the same file as the composite type if file information available */
+		tags = find_scope_members_tags(tag->file ? tag->file->tags_array : tags_array, tag, namespace);
 
-	g_return_val_if_fail (name && name[0] != '\0', NULL);
+	return tags;
+}
 
-	if (!tags)
-		tags = g_ptr_array_new ();
 
-	while (1)
+/* Checks whether member with member_scope directly accessible from method with
+ * method_scope */
+static gboolean member_at_method_scope(const gchar *method_scope, const gchar *member_scope,
+	langType lang)
+{
+	const gchar *sep = scope_sep(lang);
+	gboolean ret = FALSE;
+	gchar **comps;
+	guint len;
+
+	/* method scope is in the form ...::class_name::method_name */
+	comps = g_strsplit (method_scope, sep, 0);
+	len = g_strv_length(comps);
+	if (len > 1)
 	{
-		const GPtrArray *tags2;
-		guint got = 0;
-		TMTagType types = (tm_tag_class_t
-						   tm_tag_struct_t | tm_tag_typedef_t |
-						   tm_tag_union_t | tm_tag_enum_t);
+		gchar *method, *inner_scope, *cls, *cls_scope;
+		GPtrArray *tags;
+
+		/* get method/member scope */
+		method = comps[len - 1];
+		comps[len - 1] = NULL;
+		inner_scope = g_strjoinv(sep, comps);
+		comps[len - 1] = method;
+
+		/* get class scope */
+		cls = comps[len - 2];
+		comps[len - 2] = NULL;
+		cls_scope = g_strjoinv(sep, comps);
+		comps[len - 2] = cls;
+
+		/* check whether the class exists */
+		tags = tm_workspace_find(cls, cls_scope, tm_tag_class_t | tm_tag_struct_t, NULL, FALSE, lang);
+		ret = tags->len > 0;
```

Comment:
and there's no need to assign here as it's either already FALSE or the strcmp() call below will change it :)